### ISSUE TITLE
libunbound: couldn't open library / libunbound not correctly initialized

### ISSUE THREAD

**Issue by nataedu** [lieser/dkim_verifier#170]
I updated my Thunderbird to the last version and the DKIM verificator add-on stopped working. The status bar shows this brief error : Error in the DNS resolver.

If I open the error console setting the debeug levels I have the following errors : 
```
DKIM_Verifier.libunbound	DEBUG	libunboundWorker: Posting exception back to main script: Error: libunbound not correctly initialized (ub_resolve missing)
libunbound.jsm:297
DKIM_Verifier.Verifier	ERROR	DKIM verifier internal error (DKIM_DNSERROR_UNKNOWN): libunboundWorker: Error: libunbound not correctly initialized (ub_resolve missing)(resource://dkim_verifier/helper.jsm:430:3) JS Stack trace: libunboundWorker.onmessage@libunbound.jsm:316:16
libunbound.jsm:316:16
DKIM_Verifier.Verifier	DEBUG	Exception on DKIM-Signature 1
dkimVerifier.jsm:1295

```

**Comment by lieser** [comment]
@nataedu Could you fix the issue? If yes, please close this.

**Comment by SilmorSenedlen** [comment]
@lieser I think it's need to update wiki section about libunbound configuration.
Something like this:

1. Go to download page of Unbound DNS resolver: [https://nlnetlabs.nl/projects/unbound/download/](https://nlnetlabs.nl/projects/unbound/download/)
2. Download _"Binaries (no install)"_ zip package depending on your version of Thunderbird: 32-bit(x86) or 64-bit(x64)
3. Open downloaded archive and go to libunbound sub-directory `unbound-1.9.3.zip\libunbound\`
4. Extract following file:

For 32-bit(x86):
```
libcrypto-1_1.dll
libssl-1_1.dll
libunbound-8.dll
```
Or for 64-bit(x64):
```
libssl-1_1-x64.dll
libcrypto-1_1-x64.dll
libunbound-8.dll
```

To your Thunderbird profile extensions directory.

For example:

`C:\Users\your_profile_name\AppData\Roaming\Thunderbird\Profiles\profile_name\extensions\`

5. Open DKIM Verifier add-on options.
6. Go to General >> DNS
7. Select **libunbound** in Resolver dropdown list.
8.  Set Path as `extensions/libunbound-8.dll` and set **Path relative to profile directory** checkbox.


Tested on Thunderbird 68.1.0 (64-bit) with DKIM Verifier 3.0.1 and DLL's from Unbound 1.9.3 package.

**Comment by nataedu** [comment]
So @lieser I couldn't fix the issue, and after following step by step the instructions given by @SilmorSenedlen I could not fix the issue. I still get an error : 
`23/09/2019, 17:40:36	DKIM_Verifier.libunbound	FATAL	Exception in libunboundWorker: DKIM verifier internal error (DKIM_DNSERROR_UNKNOWN): libunboundWorker: Error: couldn't open library C:\Users\<user>\AppData\Roaming\Thunderbird\Profiles\<profile>\extensions/libunbound-8.dll: error 126(resource://dkim_verifier/helper.jsm:430:3) JS Stack trace: libunboundWorker.onmessage@libunbound.jsm:316:16`

I am also in Thunderbird 68.1.0 (64-bit) with DKIM verifier 3.0.1 and downloaded the DLL from Unbound 1.9.3 for the x64 architecture

**Comment by SilmorSenedlen** [comment]
@nataedu looking more closely to your DLL's path from the log:

`C:\Users\<user>\AppData\Roaming\Thunderbird\Profiles\<profile>\extensions/libunbound-8.dll`

on the part `\extensions/`

It's strange. In my case `extensions/libunbound-8.dll and Path relative to profile directory checkbox` work well.

Try to change '/' to '\' in the Path

**Comment by nataedu** [comment]
I know. I put the relative path as you had suggested, and tried both '\' and '/'. The file is there but I get the same error

**Comment by SilmorSenedlen** [comment]
Try to remove add-on and delete all preferences with prefix `extensions.dkim_verifier` from Options >> Advanced >> Config Editor(about:config) and restart TB.

**Comment by nataedu** [comment]
I tried and I still get the same error.

**Comment by d4k0** [comment]
@nataedu 
I have the exact same problem with Thunderbird 68.1.0 x64, DKIM verifier 3.0.1 and Unbound 1.9.3 x64. Somehow the extension can't load the DLL:

`23.9.2019, 21:04:20	DKIM_Verifier.libunbound	FATAL	Exception in libunboundWorker: DKIM Verifier Interner Fehler (DKIM_DNSERROR_UNKNOWN): libunboundWorker: Error: couldn't open library C:\Users\<user>\AppData\Roaming\Thunderbird\Profiles\wnxacr0f.default\extensions\libunbound-8.dll: error 126(resource://dkim_verifier/helper.jsm:430:3) JS Stack trace: libunboundWorker.onmessage@libunbound.jsm:316:16
libunbound.jsm:322
`

The path that is shown in the log is correct. Before the 64-bit version of Thunderbird existed, it worked with the x86 version of libunbound if I remember correctly (I responded in a ticket back then).

I will probably test it again with a portable version of Thunderbird and a new profile this weekend.

**Comment by lieser** [comment]
I did some experimenting, and found that the loading of the dependent libraries has some weird behavior. If it was one loaded by TB, TB seems to always try to load it from the same path again. Restarting TB, or even Windows did not change anything for me.

For testing I now added the ability to explicitly load more libraries before libunbound.
Please try it with [dkim_verifier@pl-beta.zip](https://github.com/lieser/dkim_verifier/files/3644451/dkim_verifier%40pl-beta.zip)

You also need to specify the path to the dependencies, separated by `;`, in the correct order.
E.g.:
```
extensions/libcrypto-1_1-x64.dll;extensions/libssl-1_1-x64.dll;extensions/libunbound-8.dll
```

If this does still not help, please enable tracing by setting `extensions.dkim_verifier.logging.console` to `Trace`, and post the `DKIM_Verifier.libunbound` messages.

Also try to load the libraries in the above order with `ctypes.open(<full path>)`.

**Comment by d4k0** [comment]
@lieser
Nice, it seems to work now!

**Comment by nataedu** [comment]
@lieser Works for me too! I have also the request from the Windows Firewall.
Thanks for the help!

**Comment by Pascal666** [comment]
> 1. Go to download page of Unbound DNS resolver: https://nlnetlabs.nl/projects/unbound/download/
> 2. Download _"Binaries (no install)"_ zip package depending on your version of Thunderbird: 32-bit(x86) or 64-bit(x64)

The DLLs appear to no longer be available on that page.  Does anyone know the new location or have a copy they can post?

**Comment by SilmorSenedlen** [comment]
> Does anyone know the new location

Sadly not. We should ask the developers of unbound.

> have a copy they can post?
Below are archives with 3 previously mentioned files from the last published binary version(1.9.3).
[libunbound_files_1.9.3_x86.zip](https://github.com/lieser/dkim_verifier/files/4291646/libunbound_files_1.9.3_x86.zip)
[libunbound_files_1.9.3_x64.zip](https://github.com/lieser/dkim_verifier/files/4291647/libunbound_files_1.9.3_x64.zip)
